core/src/main/java/com/opensymphony/xwork2/ognl/accessor/CompoundRootAccessor.java

```diff
@@ -233,24 +234,22 @@ public Object callMethod(Map context, Object target, String name, Object[] objec
 
             if ((argTypes == null) || !invalidMethods.containsKey(mc)) {
                 try {
-                    Object value = OgnlRuntime.callMethod((OgnlContext) context, o, name, objects);
-
-                    if (value != null) {
-                        return value;
-                    }
+                    return OgnlRuntime.callMethod((OgnlContext) context, o, name, objects);
                 } catch (OgnlException e) {
                     // try the next one
-                    Throwable reason = e.getReason();
+                    reason = e.getReason();
 
-                    if (!context.containsKey(OgnlValueStack.THROW_EXCEPTION_ON_FAILURE) && (mc != null) && (reason != null) && (reason.getClass() == NoSuchMethodException.class)) {
+                    if ((mc != null) && (reason != null) && (reason.getClass() == NoSuchMethodException.class)) {
                         invalidMethods.put(mc, Boolean.TRUE);
-                    } else if (reason != null) {
-                        throw new MethodFailedException(o, name, e.getReason());
```

Comment:
Was the old logic intentionally limiting the loop to 1 iteration (unless the exception reason was `null` ) ?
Previously it would seem things would _only succeed_ if the method was defined on the object at the top of the stack. 

Will "walking the whole valuestack" potentially introduce a different issue (unexpectedly) ?

Reply:
> Was the old logic intentionally limiting the loop to 1 iteration (unless the exception reason was `null` ) ?
> Previously it would seem things would _only succeed_ if the method was defined on the object at the top of the stack.

In case of `throwExceptionOnFailure=true`, Yes, but it is a wrong behavior. Please see my test at https://github.com/apache/struts/blob/d4dd3386cc02833fa7113a496092b2e09a53c9e4/core/src/test/java/com/opensymphony/xwork2/ognl/OgnlValueStackTest.java#L958 which fails on Struts 2.5 (5/5/2016) on L#975.

> Will "walking the whole valuestack" potentially introduce a different issue (unexpectedly) ?

No, it is already current behavior in case of `throwExceptionOnFailure=false`.